README.md
# MJPANCAKE Origin Pack

## 这是什么

这是一个[松饼麻雀](https://mjpancake.github.io/)的非官方脚本仓库，存放我已经制作而未见于官方发行版本的人物脚本。松饼麻雀是一个致力于比官方游戏更好地呈现 Saki 世界观的麻将游戏（自称麻将模拟器但曾经支持线上游戏）。

新上传的脚本通常只经过我（打得爽就完事了）的简单测试。如果您对角色能力表现、代码逻辑等有何高见，请通过邮件或企鹅告诉我。

## 如何使用

在「松饼麻雀」-「人物下载」中，选择 MJPANCAKE Origin Pack 并「更新人物包」。在那之后，您可以在选人模式下的「下载」中看到可选人物。

## 它报错了

这是由于角色子模块一处已经被修复的 Bug 没有被更新到游戏本体中。并非所有的角色都会涉及到这个 Bug，您可以选择暂时不使用会报错的角色，等待更新。或者，您可以参考官方指导装配 libsaki 库，自行编译到可用的版本。

## 更新说明

### Feb 28, 2019

+ [宮永咲](#宮永咲)
	+ 修复了岭上开花成功率过低的问题。

+ 由于各种各样的原因，包括但不限于松饼麻雀为 Lua 提供的接口过少、未修复的 Bug、已经修复但迟迟没有部署的 Bug、许诺但未完成的 UI 更换，最重要的是创始人近乎弃坑的情况下难以修复的服务端，这个仓库将无限期停止更新，直到创始人回坑为止。
+ 这个仓库设计的早期，我为我写出的每个人物设计适合在联机模式下运行的模式，小心翼翼地把握联机情景下的平衡性，正是因为无比期待她们有一天在真实的麻将对战中活跃。但是松饼麻雀开发者的稀少和开发的停滞逐渐令我失望。在师傅的建议下，我决定放弃亲自为松饼麻雀更换 UI 和重写服务端的「创始人都弃坑的项目最好就别搞了」的想法。
+ 我仍然期待去实现一个在线的「松饼麻雀」，而非对着三狗打单机到睡着。或许基于可得的 Web 麻将实现，或许等到有新的开发者加入时将松饼麻雀 Fork 重写。无论如何，我要暂时向这个陪伴了我后半个寒假的仓库说再见了。

## 人物列表

+ [高鴨穏乃](#高鴨穏乃)
+ [天江衣](#天江衣)
+ [妹尾佳織](#妹尾佳織)
+ [Aislinn Wishart](#aislinn-wishart)
+ [宮永咲](#宮永咲)

## 高鴨穏乃

> 踏过现实中修行的山路  
> 越过尘世间山谷的深处  
> 在那前方的  
> 是深山幽谷的化身

+ 深山之主
	+ 削弱其他三家能力对本局配牌的影响。随着局数增加，能够影响更强的能力和进行更大幅度的削弱。
	+ 使其他三家进张趋于混沌。随着局数增加和牌山深入，其他三家的摸牌结果将愈发难以预测。
	+ 使自己的进张变好。随着局数增加和牌山深入，这个能力将益发明显。

## 天江衣

> 真舒服的风呢  
> 衣喜欢这个季节  
> 不太冷也不太热  
> 新绿的光辉  对夏天的期待高涨  
> 每年的这个时期好像都在变短  真是遗憾  
> 母亲常常这么说  
> 长期的自我实现得不到幸福和快乐  
> 幸福只存在于刹那之间  
> 想分享这些的心向往着爱和娱乐  
> 但是  麻将做不到这些  
> 麻将未必能让四个人分享快乐  
> 和衣打牌的大家  都露出了是像看到了世界末日一样的表情  
> 衣因此  又独自一人被留下来了 

+ 魑魅魍魉
	+ 使其他三家在手牌达到一向听以后进展停滞。
		+ 岭上和鸣牌不受作用。
	+ 容易在早巡直击和牌。
+ 海底捞月
	+ 听牌的下一巡摸牌前将一枚和牌沉入牌山底部，当有人到达海底时取出。
		+ 可以感知和牌残枚不足的情形。
		+ 完成换听的下一巡摸牌前，重复发动此技能。
			+ 此前被沉入海底的和牌，除非受到他家能力作用，将永远沉底。
	+ 容易以鸣牌控制海底方位，相应的，到达海底方位后，他家不容易以鸣牌错开。

## 妹尾佳織

> 三个一堆，三个一堆……  
> 啊，门清自摸对对和。

+ Beginner's Luck
	+ 乱打时牌运变好。
		+ 退向听切张、不进向听切张、他家立直打生张、晚巡打生张等皆可能触发此技能。

## Aislinn Wishart

> “……”  
> （举起画板）

+ 绘制理想
	+ 每局摸第一张牌前，确定接下来 13 巡没有发生鸣牌的情形下自己的进张。
		+ 鸣牌并不会改变真实的进张顺序，他家会取代 Aislinn 获得预测的进张。
+ 快速听牌
	+ 没有被鸣牌错开的情况下，Aislinn 有较大概率会在 13 巡内听牌。

## 宮永咲

> 麻雀って、楽しいね！

+ 岭上开花
	+ 容易做出碰和杠。
	+ 存在碰副露并有杠材剩余的情况下，容易摸到杠材。
	+ 从岭上容易摸到有效牌。
+ 牌风
	+ 容易摸到西风。

History.md
# 历史更新

### Feb 28, 2019

+ [Aislinn Wishart](#aislinn-wishart)
	+ 完全重写。
	+ 现在 Aislinn 被允许独占预测进张，只有极小概率或受他家能力影响时预测出错。
	+ 增加了预测进张方位、当前巡目预测进张和下一巡目预测进张的提示。
	+ **Aislinn 在测试中偶尔出现难以理解、难以复现的崩溃，如果您在使用中遇到相同问题，请保存截图和错误报告并联系我。**
+ [宮永咲](#宮永咲)
	+ 新增角色。
	+ **宮永咲是一个未完成的角色，亟待提出有效算法和改进 API 来完善她。在此之前，只提供现有的解决方案。**

### Feb 22, 2019

[妹尾佳織](#妹尾佳織)：略微削弱了「Beginner's Luck」的部分效果。

### Feb 21, 2019

[高鴨穏乃](#高鴨穏乃)：修改了「深山之主」的算法。  
[妹尾佳織](#妹尾佳織)：完全重写。

### Feb 19, 2019

[天江衣](#天江衣)：重写了鸣牌控制代码。

### Feb 16, 2019

[天江衣](#天江衣)：暂时移除了一部分有关鸣牌控制的代码。  
新增 [Aislinn Wishart](#aislinn-wishart)。  

### Feb 15, 2019

[天江衣](#天江衣)：加强「海底捞月」。  
新增[妹尾佳織](#妹尾佳織)。

### Feb 14, 2019

[高鴨穏乃](#高鴨穏乃)：调整「深山之主」。  
新增[天江衣](#天江衣)。

### Feb 13, 2019

新增[高鴨穏乃](#高鴨穏乃)。

Saki.girl.lua
rate_self = 1
rate_others = 1
incmk_rins = 1000
incmk_3f = 40
incmk_pg = {[0] = 0, [1] = 0, [2] = 80, [3] = 160, [4] = 0}
incmk_dkg = {[0] = 0, [1] = 40, [2] = 80, [3] = 160, [4] = 0}
incmk_kakan = 10000

-- 增加西風配牌
function onmonkey()
    local exist = exists[self:index()]
    exist:incmk(T34.new("3f"), incmk_3f * rate_self)
end

function checkinit()
    if who ~= self or iter > 500 then
        return init:ct(T34.new("3f")) == 0
    end

    if init:ct(T34.new("3f")) >= 1 and init:step7() <= 3 then
        koncai = {}
        cnt_koncai = 0
        return true
    end

    return false
end

function ondraw()
    local hand_self = game:gethand(self)
    local closed_self = hand_self:closed()
    local bark_self = hand_self:barks()
    local effa_self = hand_self:effa()
    local hand_other = game:gethand(who)
    local effa_other = hand_other:effa()
    local ready_self = hand_self:ready()

    -- 沈底西風，加槓槓材
    if who ~= self then
        mount:lighta(T34.new("3f"), (-1) * incmk_3f * rate_others)
        for i = 1, cnt_koncai do
            mount:lighta(koncai[i], (-1) * incmk_kakan * rate_others)
        end
    end

    -- 容易碰/明槓
    if who ~= self and not (ready) then
        for _, t in ipairs(T34.all) do
            local flag = true

            for _, tt in ipairs(effa_other) do
                if t:id34() == tt:id34() then
                    flag = false
                    break
                end
            end

            if flag then
                -- print(t:str34() .. " pg " .. incmk_pg[closed_self:ct(t)] * rate_others)
                mount:lighta(t, incmk_pg[closed_self:ct(t)] * rate_others)
            else
                mount:lighta(t, (-1) * incmk_pg[closed_self:ct(t)] * rate_others)
            end
        end
    end

    if who == self then
        -- 容易獲得對/刻/暗槓/西風
        for _, t in ipairs(T34.all) do
            -- print(t:str34() .. " dkg" .. incmk_dkg[closed_self:ct(t)] * rate_self)
            mount:lighta(t, incmk_dkg[closed_self:ct(t)] * rate_self)
        end

        mount:lighta(T34.new("3f"), incmk_3f * rate_self)

        -- 容易加槓
        for _, m in ipairs(bark_self) do
            if m:type() == "pon" then
                -- print(m[1]:str34() .. " kakan " .. incmk_kakan * rate_self)
                mount:lighta(m[1], incmk_kakan * rate_self)
                mount:lighta(m[1], incmk_kakan * rate_self, true)
            end
        end

        -- 從嶺上獲得有效牌
        if rinshan then
            for _, t in ipairs(effa_self) do
                -- print(t:str34() .. " rins " .. incmk_rins * rate_self)
                mount:lighta(t, incmk_rins * rate_self, true)
            end
        end
    end
end

function ongameevent()
    if event.type == "barked" and event.args["who"] == self and event.args["bark"]:type() == "pon" then
        local bark = event.args["bark"]
        cnt_koncai = cnt_koncai + 1
        koncai[cnt_koncai] = bark[1]
    end
end

-- todo：正負零，開槓不加寶牌


Kaori.girl.lua
rate = 10
incmk = 40
luck_limit = 100
incluck_step = 100
incluck_stepin = 15
incluck_stranger = 40

function checkinit()
    if who ~= self then
        return true
    end

    local hand = game:gethand(self)
    step = hand:step()
    return true
end

function ondraw()
    if who ~= self then
        return
    end

    local hand = game:gethand(self)

    -- 獎勵
    if luck >= luck_limit then
        luck = luck - luck_limit

        for _, t in ipairs(hand:effa()) do
            mount:lighta(t, incmk * rate)
        end
    end
end

function ongameevent()
    -- round-started
    if event.type == "round-started" then
        luck = 0
        richied = false
        nearend = false
        action_self = false
    end

    -- drawn
    if event.type == "drawn" and event.args["who"] == self then
        action_self = true
    end

    -- barked
    if event.type == "barked" and event.args["who"] == self then
        action_self = true
    end

    -- discarded
    if event.type == "discarded" and action_self then
        action_self = false
        local hand = game:gethand(self)
        local right = self:right()
        local cross = self:cross()
        local left = self:left()
        local river0 = game:getriver(self)
        local river1 = game:getriver(right)
        local river2 = game:getriver(cross)
        local river3 = game:getriver(left)
        local stranger = true
        local swapout = river0[#river0]
        local step_loc = hand:step()

        -- 晚巡判斷
        if nearend == false and #river0 >= 9 then
            nearend = true
        end

        -- 生張判斷
        if nearend then
            for _, v in ipairs(river1) do
                if v:id34() == swapout:id34() then
                    stranger = false
                    break
                end
            end
        end

        if nearend and stranger then
            for _, v in ipairs(river2) do
                if v:id34() == swapout:id34() then
                    stranger = false
                    break
                end
            end
        end

        if nearend and stranger then
            for _, v in ipairs(river3) do
                if v:id34() == swapout:id34() then
                    stranger = false
                    break
                end
            end
        end

        -- 晚巡生張獎勵
        if nearend and stranger then
            luck = luck + incluck_stranger
        end

        -- 立直生張獎勵
        if richied and stranger then
            luck = luck + incluck_stranger
        end

        -- 不進向聽
        if step_loc == step then
            luck = luck + incluck_stepin
        end

        -- 退向聽
        if step_loc > step then
            step = step_loc
            luck = luck + incluck_step
        end

        -- 進向聽
        if step_loc < step then
            step = step_loc
        end
    end

    -- riichi-established
    if richied == false and event.type == "riichi-established" and event.args["who"] ~= self then
        richied = true
    end
end


Koromo.girl.lua
rate_self = 1 -- 对自身影响倍率
rate_other = 1 -- 对他家影响倍率
undersea = T37.new("1p") -- 海底牌

-- compare table
function compare(a, b)
    local ta = type(a)
    local tb = type(b)
    if ta ~= "table" and tb ~= "table" then
        return a == b
    end
    if ta ~= tb then
        return false
    end
    if (#ta) ~= (#tb) then
        return false
    end
    for ka, va in pairs(a) do
        if not (compare(b[ka], va)) then
            return false
        end
    end
    return true
end

-- copy table
function copy(obj)
    if type(obj) ~= "table" then
        return obj
    end
    local res = {}
    for k, v in pairs(obj) do
        res[copy(k)] = copy(v)
    end
    return res
end

-- 判斷筋牌（有序）
function isJin(t1, t2)
    return t1:isnum() and t2:isnum() and t1:suit() == t2:suit() and t1:val() + 3 == t2:val()
end

-- 判斷可以明槓
function canMinKan(closed, t)
    return closed:ct(t) == 3
end

-- 判斷可以碰
function canPon(closed, t)
    return closed:ct(t) >= 2
end

-- 判斷可以吃
function canChii(closed, t)
    return canChiiAsLeft(closed, t) or canChiiAsMiddle(closed, t) or canChiiAsRight(closed, t)
end

-- 左副露吃
function canChiiAsLeft(closed, t)
    if (t:isz() or t:val() >= 8) then
        return false
    end

    -- 這裡用 dora hack next
    local ccal_n = t:dora()
    local ccal_nn = ccal_n:dora()
    if (not (closed:ct(ccal_n) > 0 and closed:ct(ccal_nn) > 0)) then
        return false
    end

    -- 計算食替
    for _, ccal_t in ipairs(T34.all) do
        local ccal_ct = closed:ct(ccal_t)
        if
            ccal_ct >= 1 and ((not (ccal_t == ccal_n) and not (ccal_t == ccal_nn)) or ccal_ct >= 2) and
                not (ccal_t == t) and
                not (isJin(t, ccal_t))
         then
            return true
        end
    end

    return false
end

-- 中副露吃
function canChiiAsMiddle(closed, t)
    if (t:isz() or t:val() == 1 or t:val() == 9) then
        return false
    end

    -- 這裡用 indicator hack prev
    local ccal_n = t:dora()
    local ccal_p = t:indicator()
    if (not (closed:ct(ccal_n) > 0 and closed:ct(ccal_p) > 0)) then
        return false
    end

    -- 計算食替
    for _, ccal_t in ipairs(T34.all) do
        local ccal_ct = closed:ct(ccal_t)
        if ccal_ct >= 1 and ((not (ccal_t == ccal_n) and not (ccal_t == ccal_p)) or ccal_ct >= 2) and not (ccal_t == t) then
            return true
        end
    end

    return false
end

-- 右副露吃
function canChiiAsRight(closed, t)
    if (t:isz() or t:val() <= 2) then
        return false
    end

    local ccal_p = t:indicator()
    local ccal_pp = ccal_p:indicator()
    if (not (closed:ct(ccal_p) > 0 and closed:ct(ccal_pp) > 0)) then
        return false
    end

    -- 計算食替
    for _, ccal_t in ipairs(T34.all) do
        local ccal_ct = closed:ct(ccal_t)
        if
            ccal_ct >= 1 and ((not (ccal_t == ccal_p) and not (ccal_t == ccal_pp)) or ccal_ct >= 2) and
                not (ccal_t == t) and
                not (isJin(ccal_t, t))
         then
            return true
        end
    end

    return false
end

-- 每局开始时初始化全局变量
function checkinit()
    if who ~= self then
        return true
    end

    ready = false

    -- 处理天听海底
    if init:ready() then
        local succ = false -- 本次沉底是否成功
        local listen = init:effa() -- 听牌

        for _, t in ipairs(listen) do
            if mount:remaina(t) >= 1 then
                mount:loadb(t, 1)
                succ = true
                ready = true
                undersea = t
                wait = copy(listen) -- 记录听牌
                print(t:str34() .. " 已经沉入海底")
                break
            end
        end
        if not (succ) then
            print("海底残枚不足")
        end
    end

    return true
end

function ondraw()
    remain = mount:remainpii()
    local hand = game:gethand(who)

    -- 计算海底方位
    local mod = remain % 4
    local time = math.floor(remain / 4)

    -- 地狱一向听
    if who ~= self and (not (rinshan)) and hand:step() == 1 then
        for _, t in ipairs(hand:effa()) do
            local n = mount:remaina(t)
            mount:lighta(t, (n * (-10)) * rate_other)
        end
    end

    -- 非海底路线提高直击机会
    if who ~= self and game:gethand(self):ready() and not (ready) then
        for _, t in ipairs(game:gethand(self):effa()) do
            local n = mount:remaina(t)
            mount:lighta(t, (n * 10) * rate_other)
        end
    end

    -- 提高鸣牌机会
    if who == self:left() and (not (rinshan)) and mod == 1 and not (ready) then
        closed = game:gethand(self):closed()
        it_hand = hand
        for _, t in ipairs(T34.all) do
            if canChii(closed, t) then
                mount:lighta(t, 125 * rate_other)
            end
            for _, tt in ipairs(it_hand:effa()) do
                if tt == t then
                    mount:lighta(t, -125 * rate_other)
                    break
                end
            end
        end
    end

    if who ~= self and (not (rinshan)) and mod == 1 and not (ready) then
        closed = game:gethand(self):closed()
        it_hand = hand
        for _, t in ipairs(T34.all) do
            if canPon(closed, t) then
                mount:lighta(t, 125 * rate_other)
            end
            for _, tt in ipairs(it_hand:effa()) do
                if tt == t then
                    mount:lighta(t, -125 * rate_other)
                    break
                end
            end
        end
    end

    -- 準備海底後，壓制他家鳴牌
    if who ~= self and (not (rinshan)) and ready then
        closedr = game:gethand(who:right()):closed()
        closedl = game:gethand(who:left()):closed()
        closedc = game:gethand(who:cross()):closed()
        for _, t in ipairs(T34.all) do
            if canChii(closedr, t) or canPon(closedr, t) or canPon(closedl, t) or canPon(closedc, t) then
                mount:lighta(t, -125 * rate_other)
            end
        end
    end

    -- 准备海底捞月
    if who == self and not (compare(hand:effa(), wait)) and hand:ready() then
        local succ = false
        local listen = hand:effa()
        for _, t in ipairs(listen) do
            if mount:remaina(t) >= 1 then
                mount:loadb(t, 1)
                succ = true
                ready = true
                undersea = t
                wait = copy(listen)
                print(t:str34() .. " 已经沉入海底")
                break
            end
        end
        if not (succ) then
            ready = false
            print("海底残枚不足")
        end
    end

    if who == self then
        -- 加速听牌
        if not (hand:ready()) then
            for _, t in ipairs(hand:effa()) do
                local n = mount:remaina(t)
                mount:lighta(t, n * 10 * rate_self)
            end
        end

        if (mod == 1) then
            print(time .. "巡后，自家到达海底")
        elseif (mod == 2) then
            print(time .. "巡后，下家到达海底")
        elseif (mod == 3) then
            print(time .. "巡后，对家到达海底")
        elseif (mod == 0) then
            print(time .. "巡后，上家到达海底")
        end
    end

    -- 封锁海底牌
    if ready and remain ~= 1 then
        mount:lightb(undersea, (-1000) * rate_self)
    end

    -- 释放海底牌
    if ready and remain == 1 then
        mount:lightb(undersea, 1000 * rate_self)
    end
end


Shizuno.girl.lua
mnttt = 136 -- 牌山總枚數
total = 84 -- 配牌后总枚数
rate_other = 1 -- 对他家影响调整倍率
rate_self = 1 -- 对自家影响调整倍率
iter_default = 64

-- 线性同余随机数发生器
a = 1103515245
c = 12345
m = 2 ^ 32 - 1
seed = 1

function random()
    seed = (a * seed + c) % m
    return seed
end

-- 猴子阶段扰乱配牌，随局数增加幅度变大
function onmonkey()
    local exist0 = exists[self:left():index()]
    local exist1 = exists[self:right():index()]
    local exist2 = exists[self:cross():index()]
    local exist3 = exists[self:index()]
    local round = game:getround()
    for _, t in ipairs(T34.all) do
        exist0:incmk(t, random() % mnttt * round * rate_other)
        exist1:incmk(t, random() % mnttt * round * rate_other)
        exist2:incmk(t, random() % mnttt * round * rate_other)
        exist3:incmk(t, random() % mnttt * round * rate_self)
    end
end

-- 随局数增加挂力增强
function checkinit()
    local round = game:getround()
    return iter > iter_default * round
end

-- 摸牌阶段影响摸牌
function ondraw()
    local round = game:getround()

    -- 扰乱他家摸牌
    if who ~= self then
        if rinshan then
            for _, t in ipairs(T34.all) do
                local rsmk = random() % (total - mount:remainrinshan()) * round * rate_other -- 岭上摸牌挂力，随局数和开杠次数增加增强
                mount:lighta(t, rsmk, true)
                mount:lightb(t, rsmk, true)
            end
        else
            for _, t in ipairs(T34.all) do
                local incmk = random() % (total - mount:remainpii() - mount:remainrinshan()) * round * rate_other -- 一般摸牌挂力，随局数增加和牌山深入增强
                mount:lighta(t, incmk)
                mount:lightb(t, incmk)
            end
        end
        return
    end

    -- 强化自家摸牌
    local hand = game:gethand(self)
    local effas = hand:effa()
    if rinshan then
        for _, t in ipairs(effas) do
            local rsmk = random() % (total - mount:remainrinshan()) * round * rate_self
            mount:lighta(t, rsmk, true)
            mount:lightb(t, rsmk, true)
        end
    else
        for _, t in ipairs(effas) do
            local incmk = random() % (total - mount:remainpii() - mount:remainrinshan()) * round * rate_self
            mount:lighta(t, incmk)
            mount:lightb(t, incmk)
        end
    end
end


Aislinn.girl.lua
rate = 1
incmk = 100000

-- 线性同余随机数发生器
a = 1103515245
c = 12345
m = 2 ^ 32 - 1
seed = 1

function random()
    seed = (a * seed + c) % m
    return seed
end

function checkinit()
    if who ~= self or iter > 100 then
        return true
    end

    return init:step() <= 4
end

function ondraw()
    if round_started then
        local hand = {}
        hand[0] = game:gethand(self)
        hand[1] = game:gethand(self:right())
        hand[2] = game:gethand(self:cross())
        hand[3] = game:gethand(self:left())
        local effas = hand[0]:effa()
        local len = #effas -- 序列长度
        local j = random() % len + 1 -- 骰子

        drawn = (-1) * self:turnfrom(game:getdealer()) -- 抽牌统计
        turn = 0 -- 回合統計
        cntturn = 0 -- turnmax 统计
        cntpaint = {} -- cnt[paint[turn]]
        paint = {} -- 约定进张
        count = {} -- 牌数统计
        loadb = false -- have been loaded to mountb
        is_kan = false -- avoid kan change order
        round_started = false

        -- 初始化 count，cntpaint
        for _, t in ipairs(T34.all) do
            count[t:id34()] = 0
            cntpaint[t:id34()] = 0
        end

        for _, t in ipairs(T34.all) do
            for i = 0, 3 do
                count[t:id34()] = count[t:id34()] + hand[i]:ct(t)
            end
        end

        for i = 1, 13 do
            local k = 1 -- 骰子计数
            local loop = false -- 有效循环
            local flag = true

            repeat
                -- 从头循环
                if k > len then
                    k = 1
                    loop = true
                end

                -- 记录进张
                local t = effas[k]
                if k >= j or loop then
                    if count[t:id34()] < 4 then
                        paint[i] = t:id34()
                        count[t:id34()] = count[t:id34()] + 1
                        cntturn = cntturn + 1
                        cntpaint[t:id34()] = cntpaint[t:id34()] + 1
                        flag = false
                        print(i .. "巡进张为" .. t:str34()) -- .. " debug " .. cntturn .. " " .. count[t:id34()])
                        break
                    end
                end

                k = k + 1
            until k > len and loop

            if flag then
                break
            end

            j = random() % len + 1
        end

        -- loaded to b
        for i = 1, cntturn do
            mount:loadb(T37.new(paint[i]), cntpaint[paint[i]])
        end
    end

    if is_kan then
        return
    end

    if drawn == 0 and turn < cntturn then
        drawn = drawn - 4
        turn = turn + 1
        local t = T34.new(paint[turn])
        local tt = 0
        local dis = who:turnfrom(self)

        mount:lightb(t, incmk * rate)

        if dis == 0 then
            print("进张位置为自家")
        elseif dis == 1 then
            print("进张位置为下家")
        elseif dis == 2 then
            print("进张位置为对家")
        elseif dis == 3 then
            print("进张位置为上家")
        end

        if turn < cntturn then
            tt = T34.new(paint[turn + 1])
            print("本巡进张预计为" .. t:str34() .. "，下一巡进张预计为" .. tt:str34())
        else
            print("本巡进张预计为" .. t:str34())
        end
    end
end

function ongameevent()
    -- round-started
    if event.type == "round-started" then
        round_started = true
    end

    -- drawn
    if event.type == "drawn" then
        if not (is_kan) then
            drawn = drawn + 1
        else
            is_kan = false
            -- print("debug-deal-kan")
        end
    end

    -- barked
    if event.type == "barked" then
        local bark = event.args["bark"]
        if bark:type() == "daiminkan" or bark:type() == "ankan" or bark:type() == "kakan" then
            is_kan = true
            -- print("debug-found-kan")
        end
    end
end
